contracts/voting/Voting.sol

```diff
@@ -0,0 +1,362 @@
+            _cities[newCity.id] = city;
+            _regionToCities[regionId].push(newCity.id);
+            cityIds[i] = newCity.id;
+            emit CitiesAdded(regionId, cityIds);
+        }
+        _startVoting(regionId);
+    }
+
+    function changeCityVotePrice(uint256 cityId, uint256 newPrice) external override isOwner {
+        if (newPrice == 0) revert VotingErrors.IncorrectValue();
+        City memory city = _cities[cityId];
+        if (city.population == 0) revert VotingErrors.IncorrectValue();
+        _cities[cityId].votePrice = newPrice;
+
+        emit VotePriceUpdated(cityId, city.votePrice, newPrice);
+    }
+
+    function addBuilding(
+        uint256 cityId,
+        uint256 mayorId,
+        Building newBuilding
+    ) external override {
+        if(!_isGoverningPeriod(_cities[cityId].regionId)) revert VotingErrors.IncorrectPeriod();
+        if(
+            mayor.ownerOf(mayorId) != msg.sender ||
+            cityToMayor[cityId] != mayorId
+        ) revert VotingErrors.WrongMayor();
+        uint256 buildingPrice = _getBuildingPrice(newBuilding);
+        if(voucherToken.balanceOf(msg.sender) < buildingPrice) revert VotingErrors.InsufficientBalance();
+        if(_ownerToBuildings[msg.sender][cityId][newBuilding] > 0) revert VotingErrors.BuildingDuplicate();
+        _ownerToBuildings[msg.sender][cityId][newBuilding] = _seasonNumber(_cities[cityId].regionId) + 1;
+        emit BuildingAdded(newBuilding, cityId, msg.sender);
+        voucherToken.transferFrom(msg.sender, address(this), buildingPrice);
+    }
+
+    function nominate(
+        uint256 mayorId,
+        uint256 cityId,
+        uint256 votes
+    ) external override {
+        if (!_cities[cityId].active) revert VotingErrors.InactiveObject();
+        if (!_isVotingPeriod(_cities[cityId].regionId)) revert VotingErrors.IncorrectPeriod();
+        if(mayor.ownerOf(mayorId) != msg.sender) revert VotingErrors.WrongMayor();
+        // get the price of those votes
+        uint256 priceInVotes = _calculateVotesPrice(mayorId, cityId, votes);
+        // check if msg.sender has enough Token balance: balanceOf(msg.sender) >= votes
+        if (votesToken.balanceOf(msg.sender) < priceInVotes) revert VotingErrors.InsufficientBalance();
+        // save user vote info
+        _cityToNominees[cityId].push(Nominee({ mayorId: mayorId, votes: votes }));
+        _cities[cityId].bank += votes;
+        emit CandidateAdded(mayorId, cityId, votes);
+        // transfer the "votes" amount from user to reward pool
+        votesToken.transferFrom(msg.sender, address(this), priceInVotes);
+    }
+
+    function chooseWinner(uint256 regionId) external override {
+        if (!_regions[regionId].active) revert VotingErrors.InactiveObject();
+        if (_isVotingPeriod(regionId)) revert VotingErrors.IncorrectPeriod();
+        uint256[] memory citiesIds = _regionToCities[regionId];
+        uint256 winnerId = 0;
+        for (uint256 i = 0; i < citiesIds.length; i++) {
```

Comment:
Potentially may fail "out of gas". Need to test how many cities are possible for the region. If not as many as needed -> need to redesign this function to include its functionality to `claimPrize` and make `claimPrize` per city